### PR TITLE
Add oneDNN LSTM primitive support for XPU inference

### PR DESCRIPTION
**Pull request by jianyizh**
## Summary

Use `dnnl::lstm_forward` primitive on XPU for LSTM inference, replacing the per-timestep fused cell approach. The oneDNN primitive processes the entire sequence internally, eliminating O(T) kernel launches.

## Changes

- **New file `aten/src/ATen/native/mkldnn/xpu/RNN.cpp`**: Implements `lstm_onednn_xpu` using `dnnl::lstm_forward`, registered via `REGISTER_XPU_DISPATCH(lstm_mkldnn_stub, ...)`. Handles weight layout transformation (PyTorch `[4*H, I]` → oneDNN `ldgoi`), weight reorder, and scratchpad management.
- **`aten/src/ATen/native/RNN.cpp`**:
  - Extended `use_mkldnn()` to return true for XPU in inference mode (float/bf16/fp16)
  - Added packed sequence unwrap: when `batch_sizes` are uniform (common in batch=1 inference), reshape to regular 3D tensor and route to oneDNN
  - Added XPU-specific `LSTMCell` path with pre-computed input gates as fallback for training

## Correctness

- oneDNN and PyTorch use identical gate order: `i, f, c̃, o`
- oneDNN formula: `gates = W·x + U·h + B` where `B = b_ih + b_hh` (we sum before passing)
- Verified max diff < 1e-6 vs CPU reference across multiple configurations

## Performance (Intel PVC, Kokoro TTS, bidirectional LSTM, hidden=256)

| Path | LSTM latency | E2E Kokoro |
|------|-------------|------------|
| Original (per-step mm + fused cell) | ~34 ms | 1.06s |
| oneDNN LSTM primitive | ~5 ms | **0.635s (-40%)** |

## Dependencies

- Depends on intel/torch-xpu-ops#3770 for the fused cell fallback path (bias fix)
- Only affects XPU. CUDA and CPU paths unchanged.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gujinghui @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -97,16 +97,24 @@ bool use_mkldnn(const Tensor& input, TensorList params, TensorList hx) {
...
+  }
 #endif
+  // XPU: use oneDNN LSTM for inference
+  if (input.is_xpu() && !at::GradMode::is_enabled() &&
+      (input.scalar_type() == kFloat || input.scalar_type() == kBFloat16 ||
+       input.scalar_type() == kHalf) &&
+      input.numel() != 0) {
+    return true;
+  }
+  return false;
```
Comment (EikanWang) [inline aten/src/ATen/native/RNN.cpp]: It would be better to merge with the CPU branch.

File: aten/src/ATen/native/RNN.cpp [hunk 2]
```diff
@@ -761,6 +765,15 @@ struct LSTMCell : Cell<std::tuple<Tensor, Tensor>, cell_params> {
       return std::make_tuple(std::move(hy), std::move(std::get<1>(result)));
     }
 
+    if (input.is_xpu()) {
+      auto igates = pre_compute_input ? input : params.linear_ih(input);
```
Comment (EikanWang) [inline aten/src/ATen/native/RNN.cpp]: Why do the changes use `linear_ih` instead of `matmul_ih`?

File: aten/src/ATen/native/RNN.cpp [hunk 3]
```diff
@@ -761,6 +765,15 @@ struct LSTMCell : Cell<std::tuple<Tensor, Tensor>, cell_params> {
...
+      auto hgates = params.matmul_hh(hx);
+      auto result = at::_thnn_fused_lstm_cell(
+          igates, hgates, cx, {}, params.b_hh());
```
Comment (EikanWang) [inline aten/src/ATen/native/RNN.cpp]: Why do the changes use `{}` instead of `params.b_ih()`?

File: aten/src/ATen/native/mkldnn/xpu/RNN.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    bool batch_first) {
+  using namespace at::native::onednn;
+  using dt = dnnl::memory::data_type;
+  using tag = dnnl::memory::format_tag;
+
+  TORCH_CHECK(!train, "oneDNN LSTM on XPU only supports inference mode");
+  TORCH_CHECK(
+      input.scalar_type() == kFloat || input.scalar_type() == kHalf ||
+          input.scalar_type() == kBFloat16,
+      "oneDNN LSTM on XPU only supports float, half, and bfloat16");
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/xpu/RNN.cpp]: The two `TORCH_CHECK` should be replaced with `TORCH_INTERNAL_ASSERT`.

**Review by EikanWang — COMMENTED** [review]
More comments are required.

**Comment by EikanWang** [comment]
@claude please review the changes.

File: aten/src/ATen/native/RNN.cpp (hunk 3 above)
Comment (jianyizh) [reply aten/src/ATen/native/RNN.cpp]: Removed, it's used by another PR

File: aten/src/ATen/native/RNN.cpp (hunk 2 above)
Comment (jianyizh) [reply aten/src/ATen/native/RNN.cpp]: Removed, it's used by another PR

**Comment by jianyizh** [comment]
Addressed review comments:
- Changed `TORCH_CHECK` to `TORCH_INTERNAL_ASSERT` in `mkldnn/xpu/RNN.cpp`
- The `linear_ih`/`{}` comments were about code that's been moved to a separate PR (#185507)

File: aten/src/ATen/native/RNN.cpp [hunk 5]
```diff
@@ -83,6 +83,13 @@ bool use_miopen(const at::Tensor& input, const double dropout_state) {
 }
 
 bool use_mkldnn(const Tensor& input, TensorList params, TensorList hx) {
+  // XPU: oneDNN LSTM for inference
+  if (input.is_xpu()) {
```
Comment (chunyuan-w) [inline aten/src/ATen/native/RNN.cpp:91]: Do we need to put this inside `#if AT_MKLDNN_ENABLED()`?

Comment (jianyizh) [reply aten/src/ATen/native/RNN.cpp:91]: XPU always enables onednn

Comment (EikanWang) [reply aten/src/ATen/native/RNN.cpp:91]: @jianyizh , chuanyuan's comments make sense. Let's guard the logic by checking the oneDNN enabling.

Comment (jianyizh) [reply aten/src/ATen/native/RNN.cpp:91]: done

**Comment by EikanWang** [comment]
@jianyizh , pls. fix the CI failures.

**Comment by jianyizh** [comment]
Thanks for the detailed review! All actionable items have been addressed:

1. **`has_projections` guard** — already present in the packed-sequence path (line 1542)
2. **`userEnabledMkldnn()` blocking XPU** — this is intentional; the global mkldnn toggle should control XPU oneDNN paths too
3. **Merge XPU into CPU branch** — done, moved inside `#if AT_MKLDNN_ENABLED()`
4. & 5. **`linear_ih`/`{}`** — removed, belongs to a separate PR (#185507)
6. **`const_data_ptr`** — already using it
7. **`TORCH_CHECK` → `TORCH_INTERNAL_ASSERT`** — done
8. **Dead dropout code** — removed
9. **Weight reorder caching** — acknowledged as future work
10. **Line length** — fixed by clang-format

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
@claude are the failures related?

File: aten/src/ATen/native/mkldnn/xpu/RNN.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+      case kFloat:
+        return dt::f32;
+      case kBFloat16:
+        return dt::bf16;
+      case kHalf:
+        return dt::f16;
+      default:
+        TORCH_CHECK(false, "Unsupported dtype");
+    }
+  };
```
Comment (guangyey) [inline aten/src/ATen/native/mkldnn/xpu/RNN.cpp:61]: Could we reuse `get_onednn_dtype` directly?

**Review by guangyey — APPROVED** [review]

**Review by atalman — APPROVED** [review]
lgtm

**Comment by guangyey** [comment]
To catch 2.13 branch cut!
@pytorchbot merge -i
